### ISSUE TITLE
Could not transition to inspecting; timed out after waiting 10s

### ISSUE THREAD

**Issue by alexmac** [aws/amazon-ecs-agent#278]
Just saw this new (to me) ECS error while trying to run a task "Could not transition to inspecting; timed out after waiting 10s" What should I look into to debug this further?

```
2015-12-31T21:12:09Z [INFO] Saving state! module="statemanager"
2015-12-31T21:12:18Z [INFO] Error transitioning container module="TaskEngine" task="Staging-alexmac-Stable:15 arn:aws:ecs:us-west-2:699897246896:task/b2cb4ca7-6191-45ce-9d7a-d79d1b17959c, Status: (CREATED->RUNNING) Containers: [JdiDB (CREATED->RUNNING),dataops_site_db (CREATED->RUNNING),WebMemcache (CREATED->RUNNING),JdiMemcache (CREATED->RUNNING),dataops_api_db (CREATED->RUNNING),WebDB (CREATED->RUNNING),CubesDB (CREATED->RUNNING),MfsDB (CREATED->RUNNING),]" container="JdiDB(repos.fbn.internal/jdi_db:tested-master) (CREATED->RUNNING)" state="RUNNING"
2015-12-31T21:12:18Z [WARN] Error with docker; stopping container module="TaskEngine" task="Staging-alexmac-Stable:15 arn:aws:ecs:us-west-2:699897246896:task/b2cb4ca7-6191-45ce-9d7a-d79d1b17959c, Status: (CREATED->RUNNING) Containers: [JdiDB (RUNNING->RUNNING),dataops_site_db (CREATED->RUNNING),WebMemcache (CREATED->RUNNING),JdiMemcache (CREATED->RUNNING),dataops_api_db (CREATED->RUNNING),WebDB (CREATED->RUNNING),CubesDB (CREATED->RUNNING),MfsDB (CREATED->RUNNING),]" container="JdiDB(repos.fbn.internal/jdi_db:tested-master) (RUNNING->RUNNING)" err="Could not transition to inspecting; timed out after waiting 10s"
```

Running self-compiled versions of docker 1.9.1 and ECS agent 1.7.0

docker info:
```
Containers: 33
Images: 91
Server Version: 1.9.1-fbn
Storage Driver: devicemapper
 Pool Name: docker_vg-docker_lv
 Pool Blocksize: 65.54 kB
 Base Device Size: 107.4 GB
 Backing Filesystem: xfs
 Data file: 
 Metadata file: 
 Data Space Used: 3.095 GB
 Data Space Total: 51.44 GB
 Data Space Available: 48.34 GB
 Metadata Space Used: 9.49 MB
 Metadata Space Total: 50.33 MB
 Metadata Space Available: 40.84 MB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Library Version: 1.02.93-RHEL7 (2015-01-28)
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.13-19.30.amzn1.x86_64
Operating System: Amazon Linux AMI 2015.09
CPUs: 2
Total Memory: 3.862 GiB
Name: ip-172-31-14-41
ID: EMOO:K4KT:NF2T:ZRS7:YXNG:VKWB:2MGA:HA2Z:6EML:67TN:JUMX:UPRA
```

**Comment by samuelkarp** [comment]
That error indicates that the Agent tried to call the equivalent of `docker inspect` and timed out while waiting for a response.  Can you enable debug logging on the Docker daemon and look to see if there's any interesting output in the daemon log?

(Just for curiosity: any particular reason why you have self-compiled versions of Docker and the Agent?)

**Comment by alexmac** [comment]
Just enabled debugging and reproed the problem, but nothing from the logs immediately jumps out, I can send it to you privately if you'd like it (Not sure how much stuff is exposed in these logs)

I switched to using custom builds of docker because old builds were really unstable, and I wanted a fix for the OOM issue in the ECS agent before you release 1.7.0. I'll probably switch back to official builds at some point, but its convenient to have the hooks in our AMI generation scripts to switch over to these custom builds when we hit bugs.

**Comment by samuelkarp** [comment]
If you'd prefer to send them privately, you can email them to skarp (at) amazon.com.

**Comment by samuelkarp** [comment]
Thanks for sharing log files privately.

Here’s the timeline I’ve reconstructed for one of the affected containers (correlating across the ECS Agent log, ECS Agent state file, and Docker daemon log):
```
2015-12-31T22:24:45Z created
2015-12-31T22:24:47Z inspect called
2015-12-31T22:24:50Z inspect called
2015-12-31T22:24:54Z start container
2015-12-31T22:24:55Z inspect called
2015-12-31T22:25:05Z inspect timeout
2015-12-31T22:25:06Z initiated stop
2015-12-31T22:25:36Z failed to exit, docker sends SIGKILL
```
This looks like I would expect given the symptom you reported, but this is the first I can remember hearing about the inspect timeout being problematic.    I do see lots of devicemapper noise in the daemon log, however, and I notice that you have a non-default devicemapper setup; I’m wondering if this is impacting the daemon enough to cause it to block while the container is starting.  Can you share details about how you’re starting the daemon?  If you were to use the Amazon-provided init scripts, the relevant settings would be located in `/etc/sysconfig/docker` and `/etc/sysconfig/docker-storage`; I’m not sure where they’ll be if you’re starting it differently.

**Comment by alexmac** [comment]
After running into docker crashes in older versions I switched to docker 1.9 (back when Amazon linux was still on 1.7). By default Amazon linux seems to use the loopback storage mode which results in this scary message in the logs:
```
time="2015-12-31T23:25:35.857393647Z" level=warning msg="Usage of loopback devices is strongly discouraged for production use. Please use `--storage-opt dm.thinpooldev` or use `man docker` to refer to dm.thinpooldev section."
```

So I googled a bit and switched over to what seemed to be encouraged by the docker team (exact commands stolen mostly from here: https://github.com/projectatomic/docker-storage-setup). I ended up with this little script to start docker:
```
#!/bin/bash

service docker stop
sleep 2

rm -f /var/log/docker
rm -rf /var/lib/docker
mkdir -p /var/lib/docker

# Create the LVM thin volume for docker to use
pvcreate /dev/xvdb
vgcreate docker_vg /dev/sdb
lvcreate -l 25%FREE -n docker_misc_lv docker_vg
lvcreate --thin -l 100%FREE -n docker_lv docker_vg

# Make sure the dev mapper entry is available
while [ ! -h /dev/mapper/docker_vg-docker_lv ] ;
do
	echo "Waiting for docker_vg-docker_lv..."
	sleep 2
done

# Keep the docker working dir off the root volume
mkfs.xfs /dev/docker_vg/docker_misc_lv
mount /dev/docker_vg/docker_misc_lv /var/lib/docker
```

And the following switch inside OPTIONS in /etc/sysconfig/docker: `--storage-opt dm.thinpooldev=/dev/mapper/docker_vg-docker_lv`


I'll see what happens if I switch back to using the loopback mode

**Comment by alexmac** [comment]
Switching to loopback does indeed resolve my issue, perhaps I should be filing a bug about this with the docker team?

**Comment by samuelkarp** [comment]
That looks pretty reasonable.  Loopback mode with devicemapper seems to work fine for Docker 1.7.1, but not particularly well for Docker 1.9+ (see my [comment](https://github.com/aws/amazon-ecs-agent/issues/238#issuecomment-165593139) for various choices, including the direct-lvm mode you're using).  Do you see this behavior with other task definitions as well, including ones that have fewer containers?  We may just need to increase the inspect timeout here to account for Docker getting slower.

**Comment by alexmac** [comment]
This is definitely the largest task def (8 containers) and the first time I've seen this error occur, I'll try fiddling with some of the other storage options and also increasing the timeout specified in docker_container_engine.go to see what happens

**Comment by alexmac** [comment]
Increasing the timeout value in the ecs agent did work around it, but I've also switched away from xfs after noticing some scary messages in /var/log/messages when using it:

```
INFO: task docker:49341 blocked for more than 120 seconds.
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.209838]       Tainted: G            E   4.1.13-19.30.amzn1.x86_64 #1
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.213748] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.218095] docker          D ffff8827870dfd48     0 49341      1 0x00000000
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.222159]  ffff8827870dfd48 ffff8813b2a44c80 ffff8827a5f38000 0000000000000246
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.226449]  ffff8827870e0000 ffff882799081c80 ffff8812f6f132c0 ffff8812f6f132e8
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.230692]  ffff8812f6f13290 ffff8827870dfd68 ffffffff814dd6a7 ffff882799081c80
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.234999] Call Trace:
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.236360]  [<ffffffff814dd6a7>] schedule+0x37/0x90
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.239186]  [<ffffffffa086bb79>] xfs_ail_push_all_sync+0xa9/0xe0 [xfs]
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.242779]  [<ffffffff810aaf20>] ? prepare_to_wait_event+0x110/0x110
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.246098]  [<ffffffffa08572e9>] xfs_unmountfs+0x59/0x170 [xfs]
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.249108]  [<ffffffffa0857dab>] ? xfs_mru_cache_destroy+0x6b/0x90 [xfs]
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.252776]  [<ffffffffa0859d36>] xfs_fs_put_super+0x36/0x90 [xfs]
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.256060]  [<ffffffff811cb666>] generic_shutdown_super+0x76/0x100
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.259337]  [<ffffffff811cba07>] kill_block_super+0x27/0x70
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.262294]  [<ffffffff811cbd19>] deactivate_locked_super+0x49/0x80
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.265251]  [<ffffffff811cc30e>] deactivate_super+0x4e/0x70
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.267948]  [<ffffffff811e85c3>] cleanup_mnt+0x43/0x90
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.270745]  [<ffffffff811e8662>] __cleanup_mnt+0x12/0x20
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.273517]  [<ffffffff81085af7>] task_work_run+0xb7/0xf0
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.276231]  [<ffffffff81013bb1>] do_notify_resume+0x51/0x80
Jan  7 20:55:31 ip-172-31-6-114 kernel: [3542627.278988]  [<ffffffff814e187c>] int_signal+0x12/0x17
```

Not sure if you want to increase the timeout in mainline, or just close this issue as its really a docker performance bug?

**Comment by samuelkarp** [comment]
What did you ultimately end up using as the timeout value?  I think we could entertain a change to that, since we'd like the ECS Agent to work under as many configurations as possible.

Regarding xfs, I know that Docker has switched to xfs by default when using devicemapper.  Do those problems go away when you use something else (ext4?)?

**Comment by euank** [comment]
We couldn't trivially reproduce the timeout you're experiencing, but I went ahead and increased the timeout. If you've got a better value or more information on how to reproduce it, we can revisit that.

I'm hopeful that https://github.com/docker/docker/issues/18314's resolution (marked for 1.10) will provide a better resolution to this.

**Comment by euank** [comment]
Tweaking the timeout has been released as part of Agent v1.7.1. We also released a new AMI with Docker 1.9.1 and a default configuration that should show improved performance (for some definition of..).

If you continue to run into issues or have new information, please do reopen this or a new issue.
For now, I don't think we have any additional action items for this issue on our end.

Best,
Euan
